### PR TITLE
Handle 'operation timed out error' during long Redshift unloads

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
It turns out, when you spend four hours or so waiting for Redshift to finish an UNLOAD command, lots of stuff can happen on the network side.

One of the symptoms is an exception (details below): `psycopg2.DatabaseError: SSL SYSCALL error: Operation timed out`

When this happens, we want to stay the course and wait for the UNLOAD to complete - Redshift doesn't quit at this point and will continue on until it's either done or you terminate the query explicitly. 

This PR adds logic to detect this exception and to wait until the unload is complete (by watching the S3 bucket for the expected files to appear, per the manifest file which Redshift writes out).

```sh
(records-mover-3.8.5) ]0;bigbookprobroz@bigbookpro:~/src/records-mover$ mvrec table2recordsdir redshift schema table s3://bucket/folder/
08:02:27 - Starting...
08:02:37 - Mover: copying from TableRecordsSource(redshift) to DirectoryFromUrlRecordsTarget by writing to s3 records directory
08:02:37 - Starting Redshift unload to s3://bucketscratch/folder/...
09:57:05 - 
Traceback (most recent call last):
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.DatabaseError: SSL SYSCALL error: Operation timed out


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/broz/src/records-mover/records_mover/records/job/mover.py", line 37, in run_records_mover_job
    return records.move(source, target, processing_instructions)
  File "/Users/broz/src/records-mover/records_mover/records/mover.py", line 101, in move
    out = records_source.\
  File "/Users/broz/src/records-mover/records_mover/records/sources/table.py", line 121, in move_to_records_directory
    records_directory.finalize_manifest()
  File "/Users/broz/src/records-mover/records_mover/db/redshift/unloader.py", line 98, in unload
    return self.unload_to_s3_directory(schema, table, unload_plan, s3_directory)
  File "/Users/broz/src/records-mover/records_mover/db/redshift/unloader.py", line 82, in unload_to_s3_directory
    out = self.db.execute(text("SELECT pg_last_unload_count()"))
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2235, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1124, in _execute_clauseelement
    ret = self._execute_context(
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
    util.raise_(
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/Users/broz/.pyenv/versions/3.8.5/envs/records-mover-3.8.5/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DatabaseError: (psycopg2.DatabaseError) SSL SYSCALL error: Operation timed out

[SQL: 
       UNLOAD (%(select)s) TO %(unload_location)s
       CREDENTIALS %(credentials)s
       MANIFEST
       
       
       DELIMITER AS %(delimiter)s
       
       
       GZIP
       
       
       ESCAPE
       
       
       
       
    ]
[parameters: {'select': 'SELECT * FROM schema.table', 'unload_location': 's3://bucket/folder/', 'credentials': 'aws_access_key_id=key_id;aws_secret_access_key=****************************************', 'delimiter': ','}]
(Background on this error at: http://sqlalche.me/e/13/4xp6)
(records-mover-3.8.5) ]0;bigbookprobroz@bigbookpro:~/src/records-mover$
```

This PR is structured as a merge into #113 for ease of diff until that PR is merged.

### CODE REVIEW

File: records_mover/db/redshift/unloader.py [hunk 1]
```diff
@@ -80,7 +80,22 @@ def unload_to_s3_directory(self,
...
+            except sqlalchemy.exc.DatabaseError as e:
+                if 'Operation timed out' in str(e):
+                    # Large database UNLOADs can take hours, and it's
+                    # likely the connection between the client and
+                    # server will get disconnected.  In this
+                    # circumstance, we want to wait until unload is
+                    # complete and then proceed.
+                    logger.info("Server disconnected - awaiting completion "
+                                "of work by checking S3 bucket...")
+                    directory.await_completion(log_level=logging.DEBUG,
```
Comment (vinceatbluelabs) [inline records_mover/db/redshift/unloader.py:94]: We want to not log every time we poll, as the logging would get pretty silly after four hours.

File: records_mover/db/redshift/unloader.py [hunk 2]
```diff
@@ -80,7 +80,22 @@ def unload_to_s3_directory(self,
...
+                                               ms_between_polls=5000,
```
Comment (vinceatbluelabs) [inline records_mover/db/redshift/unloader.py:95]: Five seconds ought to be plenty - short enough that it doesn't slow smaller moves much, long enough that it doesn't spam the AWS API too much.

File: records_mover/url/base.py [hunk 3]
```diff
@@ -201,7 +202,9 @@ def open(self, mode: str = "rb") -> IO[Any]:
         "exist in the directory."
         raise NotImplementedError()
 
-    def wait_to_exist(self) -> None:
+    def wait_to_exist(self,
+                      log_level: int = logging.INFO,
+                      ms_between_polls: int = 50) -> None:
```
Comment (vinceatbluelabs) [inline records_mover/url/base.py:207]: Let's take this existing method and make it more configurable so we get the UX and AWS API friendliness we want.

Comment (crvena-sonja) [reply records_mover/url/base.py:207]: Love a UX update, very philosophical function name here.

**Review by crvena-sonja — APPROVED** [review]
This is GREAT! :shipit:

File: tests/unit/db/bigquery/test_bigquery_loader.py [hunk 4]
```diff
@@ -229,3 +229,19 @@ def test_temporary_gcs_directory_loc(self):
...
+                                         gcs_temp_base_loc=mock_gcs_temp_base_loc)
+        self.assertTrue(bigquery_loader.has_temporary_loadable_directory_loc())
+
+    def test_temporary_loadable_directory_scheme(self):
+        mock_db = Mock(name='db')
+        mock_url_resolver = Mock(name='url_resolver')
+        mock_gcs_temp_base_loc = MagicMock(name='gcs_temp_base_loc')
+        bigquery_loader = BigQueryLoader(db=mock_db, url_resolver=mock_url_resolver,
+                                         gcs_temp_base_loc=mock_gcs_temp_base_loc)
+        self.assertEqual('gs', bigquery_loader.temporary_loadable_directory_scheme())
```
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_loader.py:247]: W292 no newline at end of file

**Pull request merged**